Sources/Spezi/Capabilities/Notifications/NotificationHandler.swift

```diff
@@ -0,0 +1,89 @@
+    ///
+    /// For more information refer to
+    /// [Handle notifications while your app runs in the foreground](https://developer.apple.com/documentation/usernotifications/handling-notifications-and-notification-related-actions#Handle-notifications-while-your-app-runs-in-the-foreground)
+    /// and [`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/usernotificationcenter(_:willpresent:withcompletionhandler:)).
+    ///
+    /// - Parameter notification: The notification that is about to be delivered.
+    /// - Returns: The option for notifying the user. Use `[]` to silence the notification.
+    func receiveIncomingNotification(_ notification: UNNotification) async -> UNNotificationPresentationOptions
+
+#if !os(macOS)
+    /// Handle remote notification when the app is running in background.
+    ///
+    /// This method is called when there is a remote notification arriving while the app is running in background.
+    /// You can use this method to download additional content.
+    ///
+    /// For more information refer to
+    /// [`application(_:didReceiveRemoteNotification:fetchCompletionHandler:)`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application)
+    /// or [`didReceiveRemoteNotification(_:fetchCompletionHandler:)`](https://developer.apple.com/documentation/watchkit/wkextensiondelegate/3152235-didreceiveremotenotification).
+    ///
+    /// - Note: The signature for this method on macOS is slightly different. It is not `async` and doesn't have a return value.
+    ///
+    /// - Parameter remoteNotification: The data of the notification payload.
+    /// - Returns: Return the respective ``BackgroundFetchResult``.
+    func receiveRemoteNotification(_ remoteNotification: [AnyHashable: Any]) async -> BackgroundFetchResult
+#else
+    /// Handle remote notification when the app is running in background.
+    ///
+    /// This method is called when there is a remote notification arriving while the app is running in background.
+    /// You can use this method to download additional content.
+    ///
+    /// For more information refer to
+    /// [`application(_:didReceiveRemoteNotification:)`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428430-application).
+    ///
+    /// - Parameter remoteNotification: The data of the notification payload.
+    func receiveRemoteNotification(_ remoteNotification: [AnyHashable: Any])
+#endif
+}
+
+
+extension NotificationHandler {
+#if !os(tvOS)
+    /// Empty default implementation.
+    func handleNotificationAction(_ response: UNNotificationResponse) async {}
+#endif
+
+    /// Empty default implementation.
+    func receiveIncomingNotification(_ notification: UNNotification) async -> UNNotificationPresentationOptions {
+        // TODO: is there are better default?
```

Comment:
The problem currently is that the default behavior basically overrides any other implementations as we just form a union of all NotificationHandler (assuming there could be multiple modules conforming to that protocol). 

So, maybe it makes sense to make the return type an optional (or a custom type with better wording?) that can signify if a implementation is not interested in specifying any options on how a notification is presented.

Reply:
I like the idea of a custom option set that makes it easier to differentiate between the "default" case where we would just assume that Spezi picks a default and a more opinionated case where the module explicitly tries to provide that information that should not be overwritten (only merged in a union with all other opinions).

What would be our default case in case no module as an option? Just everything?

Reply:
Not sure if we should create a custom OptionSet? Trying to not reinvent the wheel and stay compatible with future versions? 

The problem with the default implementation currently is how we merge multiple NotificationHandlers right now, see https://github.com/StanfordSpezi/Spezi/blob/c37f67522b3397827addb51dc965264138087998/Sources/Spezi/Spezi/SpeziNotificationCenterDelegate.swift#L49-L52

As we currently just form a union, the default implementation basically just overrides anyone else that specifies a more limited set. I would like a behavior where we form the union of implementations that actually force restrictions but use the default `all` if no implementation set a restriction.

So might we just return `UNNotificationPresentationOptions?`, where the default implementation just returns nil?

Reply:
Makes sense; thank you for the context!

Sounds logical and making the return optional is probably the easiest way to achieve this; I would be ok with this 👍